source/_posts/2016-12-29-testing-against-external-services.markdown

```diff
@@ -0,0 +1,91 @@
+
+External responses are supposed to be completely mocked out in testing, not make actual calls to an API or server. Tests exist to make sure OUR code works as expected, not to verify that no errors are raised when making external requests. A test failing because an external response was incorrect is a broken test already.
+
+##### Reframing our point of view
+
+Here’s an example to illustrate my point. Let’s say we have an object whose job is to get a list of files from a specific folder on an FTP server. For brevity’s sake, let’s define it like this with the nitty gritty details fleshed out elsewhere:
+
+    class FtpThing
+      def getListOfFiles(username, password, folderName)
+        @someConnection.open(username, password)
+        listOfFiles = @someConnection.list(folderName)
+        @someConnection.close
+        listOfFiles
+      end
+    end
+
+Now, one option for testing this is the way we did it before, something like this:
+
+    let(:ftp_thing) { FtpThing.new }
+    let(:list) { [‘file1.txt’, ‘file2.txt'] }
+
+    before do
+      # spin up the “mock" (real) server
+      # put a (real) file or two in a folder on the server
+    end
+
+    it ‘gives me the right list of files’ do
+      expect(ftp_thing.getListOfFiles(‘username’, ‘password’, ‘folderName’)).to eq(list)
+      ftpThing.getListOfFiles(‘username’, ‘password’, ‘folderName’)
+    end
+
+    after do
+      # shut down the server
+    end
+
+On the surface, this doesn’t look terrible. It’s testing the value returned from `getListOfFiles`, right? Yay that’s what we want! But… not quite. We’ve got a (real) server opened, and we’re hoping that for whatever reason our local server doesn’t fail or raise any errors and that’s how our test will pass. Instead, we need to allow a mocked response and set up our expectation this way:
+
+    let(:mock_connection) { double(‘ftp’) }
+    let(:ftp_thing) { FtpThing.new }
+    let(:list) { [‘file1.txt’, ‘file2.txt'] }
+
+    before do
+      ftp_thing.set_instance_variable(:@someConnection, mock_connection)
+      allow(mock_connection).to receive(:open)
+      allow(mock_connection).to receive(:list).and_return(list)
+      allow(mock_connection).to receive(:close)
+    end
+
+    it ‘gives me the right list of files’ do
+      expect(ftp_thing.getListOfFiles(‘username’, ‘password’, ‘folderName’)).to eq(list)
+      ftpThing.getListOfFiles(‘username’, ‘password’, ‘folderName’)
+    end
+
+See the difference? We aren’t starting a server or making a real external call—we’re just pretending that it happened and gave us the right response. The true test is whether *our* code is giving us the right value back after it makes an external call. This is also an easy way to test error responses, too! I can allow my connection to raise an error, and then test my error handling from there, rather than trying to fake out a whole server thing.
```

Comment:
@alexaltair it actually was an em-dash before I copypasta'd from Evernote. I changed it to a semicolon since that seemed more appropriate. :)